subsys/zbus/zbus.c

```diff
@@ -191,21 +196,142 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
 	return last_error;
 }
 
+static inline void _zbus_chan_update_hop(const struct zbus_channel *chan)
+{
+
+	struct zbus_channel_observation *observation;
+	struct zbus_channel_observation_mask *observation_mask;
+
+	int chan_highest_observer_priority = _MIN_THREAD_PRIORITY;
+
+	K_SPINLOCK(&_zbus_lock)
+	{
+
+		for (int16_t i = chan->data->observers_start_idx,
+			     limit = chan->data->observers_end_idx;
+		     i < limit; ++i) {
+			STRUCT_SECTION_GET(zbus_channel_observation, i, &observation);
+			STRUCT_SECTION_GET(zbus_channel_observation_mask, i, &observation_mask);
+
+			__ASSERT(observation != NULL, "observation must be not NULL");
+
+			const struct zbus_observer *obs = observation->obs;
+
+			if (!obs->data->enabled || observation_mask->enabled) {
+				continue;
+			}
+
+			if (chan_highest_observer_priority > obs->data->priority) {
+				chan_highest_observer_priority = obs->data->priority;
+			}
+		}
+		chan->data->highest_observer_priority = chan_highest_observer_priority;
+	}
+}
+
+static inline void _zbus_update_all_channels_hop(const struct zbus_observer *obs)
+{
+	struct zbus_channel_observation *observation;
+
+	int count;
+
+	STRUCT_SECTION_COUNT(zbus_channel_observation, &count);
+
+	for (int16_t i = 0; i < count; ++i) {
+		STRUCT_SECTION_GET(zbus_channel_observation, i, &observation);
+
+		if (obs != observation->obs) {
+			continue;
+		}
+
+		_zbus_chan_update_hop(observation->chan);
+	}
+}
+
+static inline void _zbus_obs_attach_to_thread(const struct zbus_observer *obs)
+{
+	__ASSERT(obs != NULL, "obs is required");
+
+	int current_thread_priority = k_thread_priority_get(k_current_get());
+
+	if (unlikely(obs->data->priority != current_thread_priority)) {
```

Comment:
This test wants to be inside the lock too.  Basically anything that touches the shared state in obs->data is going to want to be locked absent very careful analysis about what races are possible.

Reply:
Also style: I'm not personally a fan of the likely/unlikely macros in situations where you aren't deliberately trying to generate specific code sequences.  The overwhelming majority of devices Zephyr is deployed on don't have large branch prediction penalties, and in the modern world CPUs which do tend to have branch predictors that are smarter than rando C programmers like us anyway. :)

Reply:
> This test wants to be inside the lock too. Basically anything that touches the shared state in obs->data is going to want to be locked absent very careful analysis about what races are possible.

Only a single thread can change an observer's priority at a time, as an observer is "attached" to a thread. It is not expected for the `_zbus_obs_attach_to_thread` to be called more than once concurrently, so changing the observer priority without spinlock would be safe.

Reply:
> Only a single thread can change an observer's priority at a time

Is that enforced or just by convention?  Could an attacker arrange things such that two threads race in this block here and end up doing incorrect things?